File: shared/lib/go/utils/rate_limiter.go
package utils

import (
	"context"
	"sync"
	"time"
)

// RateLimiter implements a token bucket rate limiter.
type RateLimiter struct {
	rate       int           // Tokens per second
	burst      int           // Max tokens in bucket
	tokens     float64       // Current tokens
	lastUpdate time.Time     // Last token update time
	mu         sync.Mutex
}

// NewRateLimiter creates a new rate limiter.
// rate is the number of tokens per second.
// burst is the maximum number of tokens that can be accumulated.
func NewRateLimiter(rate, burst int) *RateLimiter {
	if burst < rate {
		burst = rate
	}

	return &RateLimiter{
		rate:       rate,
		burst:      burst,
		tokens:     float64(burst),
		lastUpdate: time.Now(),
	}
}

// Allow checks if a request can proceed without blocking.
func (rl *RateLimiter) Allow() bool {
	return rl.AllowN(1)
}

// AllowN checks if n requests can proceed without blocking.
func (rl *RateLimiter) AllowN(n int) bool {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	rl.refillTokens()

	if rl.tokens >= float64(n) {
		rl.tokens -= float64(n)
		return true
	}

	return false
}

// Wait blocks until a request can proceed or context is canceled.
func (rl *RateLimiter) Wait(ctx context.Context) error {
	return rl.WaitN(ctx, 1)
}

// WaitN blocks until n requests can proceed or context is canceled.
func (rl *RateLimiter) WaitN(ctx context.Context, n int) error {
	for {
		if rl.AllowN(n) {
			return nil
		}

		// Calculate wait time
		rl.mu.Lock()
		needed := float64(n) - rl.tokens
		waitTime := time.Duration(needed/float64(rl.rate)*1e9) * time.Nanosecond
		rl.mu.Unlock()

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(waitTime):
			// Continue to next iteration
		}
	}
}

// Reserve reserves n tokens for future use and returns a Reservation.
func (rl *RateLimiter) Reserve(n int) *Reservation {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	rl.refillTokens()

	if rl.tokens >= float64(n) {
		rl.tokens -= float64(n)
		return &Reservation{
			ok:    true,
			delay: 0,
		}
	}

	needed := float64(n) - rl.tokens
	delay := time.Duration(needed/float64(rl.rate)*1e9) * time.Nanosecond
	rl.tokens = 0

	return &Reservation{
		ok:    true,
		delay: delay,
	}
}

// refillTokens adds tokens based on elapsed time since last update.
func (rl *RateLimiter) refillTokens() {
	now := time.Now()
	elapsed := now.Sub(rl.lastUpdate)
	rl.lastUpdate = now

	tokensToAdd := elapsed.Seconds() * float64(rl.rate)
	rl.tokens = min(rl.tokens+tokensToAdd, float64(rl.burst))
}

// Tokens returns the current number of available tokens.
func (rl *RateLimiter) Tokens() float64 {
	rl.mu.Lock()
	defer rl.mu.Unlock()

	rl.refillTokens()
	return rl.tokens
}

// Limit returns the rate limit (tokens per second).
func (rl *RateLimiter) Limit() int {
	return rl.rate
}

// Burst returns the burst size.
func (rl *RateLimiter) Burst() int {
	return rl.burst
}

// Reservation represents a reservation for tokens.
type Reservation struct {
	ok    bool
	delay time.Duration
}

// OK returns whether the reservation is valid.
func (r *Reservation) OK() bool {
	return r.ok
}

// Delay returns the delay before the reservation becomes valid.
func (r *Reservation) Delay() time.Duration {
	return r.delay
}

// DelayFrom returns the delay from a specific time.
func (r *Reservation) DelayFrom(now time.Time) time.Duration {
	return r.delay
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

File: shared/lib/rust/common/src/rate_limiter.rs
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Mutex;
use tokio::time::sleep;

/// Token bucket rate limiter.
pub struct RateLimiter {
    inner: Arc<Mutex<RateLimiterInner>>,
}

struct RateLimiterInner {
    rate: f64,           // Tokens per second
    burst: f64,          // Max tokens in bucket
    tokens: f64,         // Current tokens
    last_update: Instant, // Last token update time
}

impl RateLimiter {
    /// Creates a new rate limiter.
    /// rate: tokens per second
    /// burst: maximum tokens that can be accumulated
    pub fn new(rate: usize, burst: usize) -> Self {
        let burst = burst.max(rate);

        RateLimiter {
            inner: Arc::new(Mutex::new(RateLimiterInner {
                rate: rate as f64,
                burst: burst as f64,
                tokens: burst as f64,
                last_update: Instant::now(),
            })),
        }
    }

    /// Checks if a request can proceed without blocking.
    pub async fn allow(&self) -> bool {
        self.allow_n(1).await
    }

    /// Checks if n requests can proceed without blocking.
    pub async fn allow_n(&self, n: usize) -> bool {
        let mut inner = self.inner.lock().await;
        inner.refill_tokens();

        if inner.tokens >= n as f64 {
            inner.tokens -= n as f64;
            true
        } else {
            false
        }
    }

    /// Waits until a request can proceed.
    pub async fn wait(&self) {
        self.wait_n(1).await
    }

    /// Waits until n requests can proceed.
    pub async fn wait_n(&self, n: usize) {
        loop {
            let wait_time = {
                let mut inner = self.inner.lock().await;
                inner.refill_tokens();

                if inner.tokens >= n as f64 {
                    inner.tokens -= n as f64;
                    return;
                }

                // Calculate wait time
                let needed = n as f64 - inner.tokens;
                Duration::from_secs_f64(needed / inner.rate)
            };

            sleep(wait_time).await;
        }
    }

    /// Returns the current number of available tokens.
    pub async fn tokens(&self) -> f64 {
        let mut inner = self.inner.lock().await;
        inner.refill_tokens();
        inner.tokens
    }

    /// Returns the rate limit (tokens per second).
    pub async fn limit(&self) -> usize {
        let inner = self.inner.lock().await;
        inner.rate as usize
    }

    /// Returns the burst size.
    pub async fn burst(&self) -> usize {
        let inner = self.inner.lock().await;
        inner.burst as usize
    }
}

impl RateLimiterInner {
    fn refill_tokens(&mut self) {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_update);
        self.last_update = now;

        let tokens_to_add = elapsed.as_secs_f64() * self.rate;
        self.tokens = (self.tokens + tokens_to_add).min(self.burst);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_rate_limiter_allow() {
        let limiter = RateLimiter::new(10, 10);

        // Should allow up to burst
        for _ in 0..10 {
            assert!(limiter.allow().await);
        }

        // Should deny after burst
        assert!(!limiter.allow().await);
    }

    #[tokio::test]
    async fn test_rate_limiter_refill() {
        let limiter = RateLimiter::new(10, 10);

        // Consume all tokens
        for _ in 0..10 {
            assert!(limiter.allow().await);
        }

        // Wait for refill (100ms should add 1 token at 10/sec)
        tokio::time::sleep(Duration::from_millis(100)).await;
        assert!(limiter.allow().await);
    }
}
